### DISCUSSION TITLE
Inserting a set of smaller images into a bigger one: questions about "-gravity" parameter

### DISCUSSION THREAD

**Discussion by bulbaka** [ImageMagick/ImageMagick#3458] [Consulting]
I'm placing a set of smaller images (objects, of size e.g. 32x32) within some big images (areas, of size e.g. 512x512).
There's a lot of objects (including the same objects with just different coordinates) and areas.
I use a script like this (a simplified example), that works pretty well:
```
    :: offsets to place objects within an area
    :: (here in the example there are two offset pairs to place the same object there twice) 
    set coo=+000+111 +222+333
    :: placing object
    for %%c in (%coo%) do (
    magick convert area.png object.png -gravity NorthWest -geometry %%c -composite area.png
    )
```
As you can see, the object placing is bound to the upper-left corner of the area (_-gravity NorthWest_), which is OK.
In its turn, that same gravity makes the objects themselves to be anchored by their own upper-left corners. That's not that OK because given coordinates are practically being measured by the objects' centers and therefore the offsets have to be recalculated according to the _w*h_ dimensions, e.g.:
_actual_horizontal_offset = measured_horizontal_offset - object_width / 2_
_actual_vertical_offset = measured_vertical_offset - object_heigth / 2_
I.e. here in my example, those offsets (+000+111 +222+333) have been recalculated from the measured values as follows:
[for 32x32 object's center measured coordinates 1: 16, 127 the offsets are:] _16-32/2=+000_, _127-32/2=+111_ 
[for 32x32 object's center measured coordinates 2: 238, 349 the offsets are:] _238-32/2=+222_, _349-32/2=+333_ 
Since I have 2668 such objects to place by now, these recalculations  (even semi-automated in excel) are being assumed a bit excessive : )

So, my question:
Is there a way to coordinate objects by their centers while keeping _-gravity NorthWest_ for the background?
Thank you!

**Comment by snibgo** [comment]
Why are you using "magick convert"? This limits you to v6 syntax. I suggest you remove "convert", so you get v7 syntax, and the following will work.

In your "+000+111", each number can then be replaced by a "%[fx:]" expression. For example (Windows BAT syntax):
```
magick -size 512x512 xc:Blue -size 32x32 xc:Yellow ^
  -gravity NorthWest -geometry +%%[fx:16-v.w/2]+%%[fx:127-v.h/2] -composite ^
  out.png
```
In the expression, "v.w" is the width of the second image, and "v.h" is the height.

**Reply by bulbaka** [reply]
@snibgo 
Thank you very much for your detailed reply and valuable input.
It took me a day to adapt your information and make it workable for my needs.
And now it works like a charm.
I've been especially afraid if this works in case of the negative coordinates (e.g. 002-32/2=-14), and it does work!  
This is true Image Magic!
Here below I'm putting a detailed example.
Sorry for the long text.
Still, maybe someone else finds it somehow helpful.

````
@echo off
:: Runtime %%var parsing
Setlocal EnableDelayedExpansion

:: This example script to run within the ImageMagick directory
:: Or the correct path to ImageMagick binaries must be set, e.g.
:: set app=%ProgramFiles%\ImageMagick\bin
:: path %path%;"%app%"

:: #########################################################
:: Setting area name and coordinates of the objects to place within
:: - given coordinates are real-life data
:: - values (in pixels) are measured by the objects' centers within a sample image from its upper-left corner
:: - values input in a spreadsheet to be processed in a format as follows (for further parsing purposes):  
::       - every single coordinate is formatted to represent a triad of digits  
::       - every single object coordinate pairs are concatenated to form a hexad of digits (triad pairs)
::       - hexads (triad/coordinate pairs) are separated by spaces
:: Here below are the coordinate data to place 5 image objects 25 times within the same area image
:: (totally, there're 2668 such objects I have so far)
:: #########################################################

set area=GitHub_01
set ca=016460 048498 106504 496064 502016 506292 
set ch=016136 122022 316460 444326 
set cl=116278 182266 256230 300342 410044 410160 
set cm=034154 040040 166036 326328 376458 424400 
set ct=046410 098242 274044 286370 316502 340084 344292 400140 504220 

:: Getting Area Image

magick -size 512x512 xc:Blue %area%.png

:: Batch processing Image Object 1:
:: - parsing each hexade to get back a pair of coordinate triads (e.g. the first 016460 here to be split to 016 and 460)    
:: - inserting Object 1 into the Area as long as coordinate hexads exist (6 times here)

for %%a in (%ca%) do (set coo=%%a
set x=!coo:~0,3!
set y=!coo:~-3!
magick %area%.png -size 32x32 xc:Aqua -gravity NorthWest -geometry +%%[fx:!x!-v.w/2]+%%[fx:!y!-v.h/2] -composite %area%.png
)

:: Batch inserting Image Objects 2-5 into the same Area (all in the same way as above):

for %%h in (%ch%) do (set coo=%%h
set x=!coo:~0,3!
set y=!coo:~-3!
magick %area%.png -size 32x32 xc:Honeydew -gravity NorthWest -geometry +%%[fx:!x!-v.w/2]+%%[fx:!y!-v.h/2] -composite %area%.png
)

for %%l in (%cl%) do (set coo=%%l
set x=!coo:~0,3!
set y=!coo:~-3!
magick %area%.png -size 32x32 xc:Lime -gravity NorthWest -geometry +%%[fx:!x!-v.w/2]+%%[fx:!y!-v.h/2] -composite %area%.png
)

for %%m in (%cm%) do (set coo=%%m
set x=!coo:~0,3!
set y=!coo:~-3!
magick %area%.png -size 32x32 xc:Moccasin -gravity NorthWest -geometry +%%[fx:!x!-v.w/2]+%%[fx:!y!-v.h/2] -composite %area%.png
)

for %%t in (%ct%) do (set coo=%%t
set x=!coo:~0,3!
set y=!coo:~-3!
magick %area%.png -size 32x32 xc:Tomato -gravity NorthWest -geometry +%%[fx:!x!-v.w/2]+%%[fx:!y!-v.h/2] -composite %area%.png
)
````

NB In my real-life DIY-task I deal with the prepared .dds area and object images (i.e. not making them on the go as in the example above).
In this case, the respective code (a single entry segment) would be as follows:
````
magick source_area.dds -define dds:mipmaps=0 -define dds:compression=dxt1 %area%.dds

for %%a in (%ca%) do (set coo=%%a
set x=!coo:~0,3!
set y=!coo:~-3!
magick %area%.dds ca.dds -define dds:mipmaps=0 -define dds:compression=dxt1 -gravity NorthWest -geometry +%%[fx:!x!-v.w/2]+%%[fx:!y!-v.h/2] -composite %area%.dds
)
````
[GitHub_01_IM7_geo_[x-y]_5objects_png_cmd_POST.zip](https://github.com/ImageMagick/ImageMagick/files/6227687/GitHub_01_IM7_geo_.x-y._5objects_png_cmd_POST.zip)
